File: Python/findTargetSumWays.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: findTargetSumWays.py
@Function: 目标和  动态规划
@Link: https://leetcode-cn.com/problems/target-sum/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-16
"""


class Solution:
    def findTargetSumWays(self, nums, S) -> int:
        sumAll = sum(nums)
        if S > sumAll or (S + sumAll) % 2:
            return 0
        target = (S + sumAll) // 2

        dp = [0] * (target + 1)
        dp[0] = 1

        for num in nums:
            for j in range(target, num - 1, -1):
                dp[j] = dp[j] + dp[j - num]
        return dp[-1]


# -------------------------
if __name__ == "__main__":
    nums = [1, 1, 1, 1, 1]
    target = 3

    solution = Solution()
    sum_ways = solution.findTargetSumWays(nums, target)
    print(f"The solution of this problem is {sum_ways}")

File: C++/findTargetSumWays.cpp
/**
 * @File    : findTargetSumWays.cpp
 * @Brief   : 目标和 动态规划
 * @Link    : https://leetcode-cn.com/problems/target-sum/
 * @Author  : Wei Li
 * @Date    : 2021-07-16
*/

#include <iostream>
#include <vector>

class Solution
{
public:
    int findTargetSumWays(std::vector<int> &nums, int target)
    {
        int sum = 0;
        for (int &num : nums)
        {
            sum += num;
        }
        int diff = sum - target;
        if (diff < 0 || diff % 2 != 0)
        {
            return 0;
        }
        int neg = diff / 2;
        std::vector<int> dp(neg + 1);
        dp[0] = 1;
        for (int &num : nums)
        {
            for (int j = neg; j >= num; j--)
            {
                dp[j] += dp[j - num];
            }
        }
        return dp[neg];
    }
};

// ------------------------------
int main(int argc, char **argv)
{
    std::vector<int> nums { 1, 1, 1, 1, 1 };
    int target = 3;

    Solution solution;
    auto sum_ways = solution.findTargetSumWays(nums, target);
    std::cout << "The solution of this problem is " << sum_ways << std::endl;

    return 0;
}
